File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        
        	vector<vector<int>> ans;
	int n = (int)nums.size();
	sort(nums.begin(), nums.end());

	for(int i = 0 ; i < n ; i ++){
		if(i > 0 && nums[i] == nums[i - 1])
			continue;
		int l = i+1, r = n-1;
		while(l < r){
			int sum = nums[i] + nums[l] + nums[r];
			if(sum > 0)
				r --;
			else if(sum < 0)
				l ++;
			else {
				ans.push_back({nums[i], nums[l], nums[r]});
				l ++;
				while(l < r && nums[l] == nums[l-1]){
					l +=1;
				}
			}
		}


	}

	return ans;
        
        
    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
            res =[]
            nums.sort()
            for idx, val in enumerate(nums):
                if idx > 0 and nums[idx] == nums[idx - 1]:
                    continue

                l, r = idx+1, len(nums) - 1
                while l < r:
                    t_sum = val + nums[l] + nums[r]

                    if t_sum < 0:
                        l += 1
                    elif t_sum > 0:
                        r -= 1
                    else:
                        res.append([val, nums[l], nums[r]])
                        l += 1
                        while l < r and nums[l] == nums[l - 1]:
                            l +=1
            return res

